src/Nethermind/Nethermind.JsonRpc/JsonRpcUrlCollection.cs

```diff
@@ -0,0 +1,115 @@
+            string environmentVariableUrl = Environment.GetEnvironmentVariable(NethermindUrlVariable);
+            if (!string.IsNullOrWhiteSpace(environmentVariableUrl))
+            {
+                if (Uri.TryCreate(environmentVariableUrl, UriKind.Absolute, out Uri? uri))
+                {
+                    defaultUrl.Scheme = uri.Scheme;
+                    defaultUrl.Host = uri.Host;
+                    defaultUrl.Port = !uri.IsDefaultPort ? uri.Port : defaultUrl.Port;
+                }
+                else if (_logger.IsWarn)
+                    _logger.Warn($"Environment variable '{NethermindUrlVariable}' value '{environmentVariableUrl}' is not valid JSON RPC URL, using default url : '{defaultUrl}'");
+            }
+            Add(defaultUrl.Port, defaultUrl);
+
+            if (includeWebSockets)
+            {
+                if (_jsonRpcConfig.WebSocketsPort != _jsonRpcConfig.Port)
+                {
+                    JsonRpcUrl defaultWebSocketUrl = defaultUrl.Clone() as JsonRpcUrl;
+                    defaultWebSocketUrl.Port = _jsonRpcConfig.WebSocketsPort;
+                    defaultWebSocketUrl.RpcEndpoint = RpcEndpoint.Ws;
+                    Add(defaultWebSocketUrl.Port, defaultWebSocketUrl);
+                }
+                else
+                    defaultUrl.RpcEndpoint |= RpcEndpoint.Ws;
+            }
+
+            foreach (string additionalRpcUrl in _jsonRpcConfig.AdditionalRPCUrls)
+            {
+                try
+                {
+                    JsonRpcUrl url = JsonRpcUrl.Parse(additionalRpcUrl);
+                    if (!includeWebSockets && url.RpcEndpoint.HasFlag(RpcEndpoint.Ws))
+                    {
+                        url.RpcEndpoint &= ~RpcEndpoint.Ws;
+                        if (url.RpcEndpoint == RpcEndpoint.None)
+                        {
+                            if (_logger.IsInfo)
+                                _logger.Info($"Additional JSON RPC URL '{url}' has web socket endpoint type and web sockets are not enabled; skipping...");
+                            continue;
+                        }
+                    }
+
+                    if (ContainsKey(url.Port))
+                    {
+                        if (_logger.IsInfo)
+                            _logger.Info($"Additional JSON RPC URL '{url}' wants port {url.Port}, but port already in use; skipping...");
+                        continue;
+                    }
+
+                    Add(url.Port, url);
+                }
+                catch (FormatException fe)
+                {
+                    if (_logger.IsInfo)
+                        _logger.Info($"Additional JSON RPC URL packed value '{additionalRpcUrl}' format error: {fe.Message}; skipping...");
+                    continue;
+                }
+                catch (Exception e)
```

Comment:
Still, do we want to catch any exception here? Is it really needed?

Reply:
My original thinking was to allow it to recover from any exception on a per url basis so that if something unexpected occurs for one url, it doesn't stop it from parsing the rest of the urls. I guess we should only recover & continue from expected exceptions (FormatException). I can clean that up.

Reply:
Cleaned up catch-all in latest commit